build-rust.py

```diff
@@ -0,0 +1,166 @@
+                    down an older bug to properly bisect. This value is just passed along to 'git checkout'
+                    so it can be a branch name, tag name, or hash. This will have no effect if
+                    '--rust-folder' is provided, as the script does not manipulate a repository that it
+                    does not own.
+
+                    '''),
+                    default='master',
+                    type=str)
+parser.add_argument('--show-build-commands',
+                    help=textwrap.dedent('''\
+                    By default, the script only shows the output of the comands it is running. When this option
+                    is enabled, the invocations of the build tools will be shown to help with reproducing
+                    issues outside of the script.
+
+                    '''),
+                    action='store_true')
+clone_options.add_argument('--use-good-revision',
+                           help=textwrap.dedent('''\
+                    By default, the script updates Rust to the latest tip of tree revision, which may at times be
+                    broken or not work right. With this option, it will checkout a known good revision of Rust
+                    that builds and works properly. If you use this option often, please remember to update the
+                    script as the known good revision will change. This option may work best with a matching good
+                    revision used to build LLVM by './build-llvm.py'.
+
+                           '''),
+                           action='store_const',
+                           const=GOOD_REVISION,
+                           dest='ref')
+parser.add_argument('--vendor-string',
+                    help=textwrap.dedent('''\
+                    Add this value to the Rust version string (like "rustc ... (ClangBuiltLinux)"). Useful when
+                    reverting or applying patches on top of upstream Rust to differentiate a toolchain built
+                    with this script from upstream Rust or to distinguish a toolchain built with this script
+                    from the system's Rust. Defaults to ClangBuiltLinux, can be set to an empty string to
+                    override this and have no vendor in the version string.
+
+                    '''),
+                    type=str,
+                    default='ClangBuiltLinux')
+args = parser.parse_args()
+
+# Start tracking time that the script takes
+script_start = time.time()
+
+# Folder validation
+tc_build_folder = Path(__file__).resolve().parent
+src_folder = Path(tc_build_folder, 'src')
+
+if args.build_folder:
+    build_folder = Path(args.build_folder).resolve()
+else:
+    build_folder = Path(tc_build_folder, 'build/rust')
+
+if args.llvm_install_folder:
+    llvm_install_folder = Path(args.llvm_install_folder).resolve()
+else:
+    llvm_install_folder = Path(tc_build_folder, 'build/llvm/final')
```

Comment:
Hmmm, I wonder if we want to look for an `llvm` folder above next to `build_folder.parent` instead…

Reply:
Actually, that would not work because the user's specified build folder may not have the same structure at the tc-build default. It should probably be an error for the user to specify `--build-folder` without `--llvm-install-folder`, as we can assume if they customized the build folder for `build-rust.py`, they likely did it for `build-llvm.py` and the default will not work.

Reply:
Sounds good -- added. I wonder if it should be a warning, but it does not matter much since it is not too much work for a user. I documented the behavior too.